docs/shared_library/README.md

```diff
@@ -0,0 +1,234 @@
+        :- use_module(library(clpz)).
+
+        as --> [].
+        as --> [a], as.
+
+        fact(0).
+        fact(1).
+        fact(2).
+        """
+
+    with ScryerMachine(source) as wam:
+        with ScryerMachine(source) as wam1:
+            for n, a in zip(range(3), wam.lazy_eval('phrase(as, As).')):
+                for i, fact in zip(range(3), wam1.lazy_eval('fact(Fact).')):
+                    print(n, i, fact, a)
+
+# 0 0 {'Fact': 0} {'As': ['']}
+# 0 1 {'Fact': 1} {'As': ['']}
+# 0 2 {'Fact': 2} {'As': ['']}
+# 1 0 {'Fact': 0} {'As': 'a'}
+# 1 1 {'Fact': 1} {'As': 'a'}
+# 1 2 {'Fact': 2} {'As': 'a'}
+# 2 0 {'Fact': 0} {'As': 'aa'}
+# 2 1 {'Fact': 1} {'As': 'aa'}
+# 2 2 {'Fact': 2} {'As': 'aa'}
+                                                        
+```
+
+  * Besides being limited to a single generative stream at a time, when the generative stream is finished or not longer needed, the resources used MUST be manually deallocated with `scryer_cleanup_query_generator`, or it will leave the Prolog machine in an indeteriminate state and future answers will not be accurate (if it doesn't cause a kernel panic). I advise you to treat a both scryer machines and generative queries the same way you would treat a file or other resource, and wrap it with a try/finally or other form of context manager, e.g.:
+  
+```python
+# ... snip ...
+class ScryerMachine:
+
+    def __init__(self, source=None):
+        self.source = source
+        self.m: MACHINE_PTR = None
+
+    def __enter__(self, query: str = None):
+        self.m = lib.scryer_machine_new() # initiate and store machine pointer
+        if self.source:
+            load_facts(self.m, self.source)
+        if query:
+            return self.lazy_eval(self.m, query)
+        return self
+
+    def __exit__(self, *_):
+        lib.scryer_machine_free(self.m) # deallocate resources when lexical scope ends
+        
+# ... snip ...
+```
+
+```python
+# ...snip...
+
+@contextlib.contextmanager
+def query_generator(machine: MACHINE_PTR, query: str):
+    qs: QUERY_STATE_PTR = lib.scryer_start_new_query_generator(machine, query.encode('utf-8'))
+    try:
+        def generator():
+            seen_answer = False
+            while True:
+                answer = run_generator_step(machine, qs)
+                if not isinstance(answer, list):
+                    if seen_answer:
+                        break
+                    yield answer[0] if isinstance(answer, list) else answer
+                    break
+                seen_answer = True
+                yield answer[0] if isinstance(answer, list) else answer
+
+        yield generator()
+    finally:
+        handle_loader_deloader(lib.scryer_cleanup_query_generator(machine, qs))
+# ...snip...
+```
+
+See [`scryer_python.py`](scryer_python.py) for a complete example.
+  
+  * Note that the JSON printer returns results that are not completely accurate when it comes to differentiating strings from symbols, e.g.:
```

Comment:
Is this still the case? I think the JSON printer implemented by @bakaq does this correctly!